spark-common/src/main/java/me/lucko/spark/common/sampler/async/TickedAsyncDataAggregator.java

```diff
@@ -0,0 +1,115 @@
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package me.lucko.spark.common.sampler.async;
+
+import me.lucko.spark.common.sampler.ThreadGrouper;
+import me.lucko.spark.common.sampler.window.WindowStatisticsCollector;
+import me.lucko.spark.common.tick.TickReporter;
+import me.lucko.spark.proto.SparkSamplerProtos;
+
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.TimeUnit;
+
+public class TickedAsyncDataAggregator extends AsyncDataAggregator {
+
+    /** The ticks that exceeded the threshold, cleared one-by-one when inserting data */
+    private final Queue<ExceededTick> ticksOver = new ConcurrentLinkedQueue<>();
+
+    /** The callback called when this aggregator is closed, to clean up resources */
+    private final Runnable closeCallback;
+
+    /** Tick durations under this threshold will not be inserted, measured in microseconds */
+    private final long tickLengthThreshold;
+
+    /** Counts the number of ticks aggregated */
+    private WindowStatisticsCollector.ExplicitTickCounter tickCounter;
+
+    protected TickedAsyncDataAggregator(ThreadGrouper threadGrouper, boolean ignoreSleeping, TickReporter tickReporter, int tickLengthThreshold) {
+        super(threadGrouper, ignoreSleeping);
+        this.tickLengthThreshold = TimeUnit.MILLISECONDS.toMicros(tickLengthThreshold);
+        TickReporter.Callback tickReporterCallback = duration -> {
+            if (duration > tickLengthThreshold) {
+                long end = System.nanoTime();
+                long start = (long) (end - (duration * 1_000_000)); // ms to ns
+                this.ticksOver.add(new ExceededTick(start, end));
+                this.tickCounter.increment();
+            }
+        };
+        tickReporter.addCallback(tickReporterCallback);
+        this.closeCallback = () -> tickReporter.removeCallback(tickReporterCallback);
+    }
+
+    @Override
+    public void insertData(ProfileSegment element, int window) {
+        // with async-profiler clock=monotonic, the event time uses the same clock
+        // as System.nanoTime(), so we can compare it directly
+        long time = element.getTime();
+        while (true) {
+            ExceededTick earliestExceeding = ticksOver.peek();
+            if (earliestExceeding == null) {
+                // no tick over threshold anymore
+                return;
+            } else if (time - earliestExceeding.start < 0) {
+                // segment happened before current exceeding
+                return;
+            } else if (earliestExceeding.end - time < 0) {
+                // segment happened after current exceeding,
+                // but it might fall into the next one
+                ticksOver.remove();
+            } else {
+                // segment falls exactly into exceeding, record it
+                break;
+            }
```

Comment:
would it be possible to add some unit tests for this logic? either by directly testing this method, or extracting the ExceededTick logic out into another class and testing it there.

mostly interested in the following cases I think:

- no ExceededTick present when aggregating
- ExceededTick added after the aggregation has happened
- ExceededTick added before the aggregation has happened
- ExceededTicks added in the wrong / non-std order (?)
- profile segments aggregated in the wrong / non-std order (?)

Reply:
I extracted the logic and added a few test cases. Let me know what you think.

Regarding ExceededTicks added in the wrong order, this would require the server to call the `onTick` method with totally wrong durations (we calculate the tick start time from the current nano time and the duration).

The profile segments should be passed in the right order by JfrReader as we use `readAllEvents` which sorts the events by time.

Reply:
Thank you - tests look good!